File: src/solutions/2023/06/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false

let ( */ ) = Int64.mul

let ohnomath (b : float) (c : float) : int64 =
  let epsilon = 0.001 in
  let mn = floor (((b +. sqrt (abs_float ((b *. b) -. (4. *. c)))) /. 2.) -. epsilon) in
  let mx = ceil (((b -. sqrt (abs_float ((b *. b) -. (4. *. c)))) /. 2.) +. epsilon) in
  Int64.of_float (mn -. mx +. 1.)
;;

let part_1 (input : string) (_is_test : bool) : string =
  let td =
    input
    |> String.split_on_char '\n'
    |> List.map (fun x ->
      x
      |> String.split_on_char ':'
      |> List.tl
      |> List.hd
      |> String.split_on_char ' '
      |> List.filter (fun s -> String.length s > 0)
      |> List.map float_of_string
      |> Array.of_list)
    |> Array.of_list
  in
  let rec loop (res : int64) (i : int) : int64 =
    if i >= Array.length td.(0)
    then res
    else loop (res */ ohnomath td.(0).(i) td.(1).(i)) (i + 1)
  in
  loop 1L 0 |> Int64.to_string
;;

let part_2 (input : string) (_is_test : bool) : string =
  let td =
    input
    |> String.split_on_char '\n'
    |> List.map (fun x ->
      x
      |> String.split_on_char ':'
      |> List.tl
      |> List.hd
      |> String.split_on_char ' '
      |> List.filter (fun s -> String.length s > 0)
      |> String.concat ""
      |> float_of_string)
    |> Array.of_list
  in
  ohnomath td.(0) td.(1) |> Int64.to_string
;;

let () = Aoc.run Sys.argv part_1 part_2 options

File: src/solutions/2023/06/python/main.py
import sys
import os
import re
from math import ceil, floor, sqrt
from typing import List

sys.path.append(os.getcwd())
from src.langs.python.run import run


def main() -> None:
   options = {'has_alternate': False, 'has_io': False}
   run(sys.argv, part_one, part_two, options)


def ohnomath(b: int, c: int) -> int:
   pepsilon = 0.001
   mn = floor((b + sqrt(abs(b * b - 4 * c))) / 2 - pepsilon)
   mx = ceil((b - sqrt(abs(b * b - 4 * c))) / 2 + pepsilon)
   return mn - mx + 1


def part_one(input: str, is_test: bool) -> str:
   td = [[int(y) for y in x.split(':')[1].strip().split(' ') if y != '']
         for x in input.split('\n')]
   res = 1
   for i in range(len(td[0])):
      res *= ohnomath(td[0][i], td[1][i])
   return str(res)


def part_two(input: str, is_test: bool) -> str:
   td = [
       ''.join(x.split(':')[1].strip().split(' ')) for x in input.split('\n')
   ]
   return str(ohnomath(int(td[0]), int(td[1])))


if __name__ == '__main__':
   main()
